### PR TITLE
feat: add struct hash

### PR DESCRIPTION
**Pull request by janek26**
simuliar to eip712

### CODE REVIEW

File: contracts/StructHash.cairo [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    member wallet : felt
+end
+
+struct Mail:
+    member from2 : Person
+    member to : Person
+    member contents : felt
+end
+
+# H('StarkNetDomain(felt name,felt version,felt chainId)')
```
Comment (juniset) [inline contracts/StructHash.cairo:27]: Given function declaration in Cairo we should do ` H('StarkNetDomain(name:felt,version:felt,chainId:felt)')` to be consistent. That's also how it is defined in https://github.com/argentlabs/argent-x/discussions/14.

And the same for the other struct definitions (Mail and Person).

File: contracts/StructHash.cairo [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+end
+
+# test view
+@view
+func test{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}() -> (hash : felt):
+    alloc_locals
+    let (__fp__, _) = get_fp_and_pc()
+
+    local domain : StarkNet_Domain = StarkNet_Domain(
+        name='Ether Mail',
```
Comment (juniset) [inline contracts/StructHash.cairo:92]: Since we are in StarkNet might as well name this `StarkNet Mail` :-)

File: contracts/StructHash.cairo [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    let (personFromHash : felt) = hashPerson(person=&mail.from2)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=personFromHash)
+    let (personToHash : felt) = hashPerson(person=&mail.to)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=personToHash)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=mail.contents)
+    let (hash : felt) = hash_finalize(hash_state_ptr=hash_state)
+    return (hash=hash)
+end
+
+func hashMessage{hash_ptr : HashBuiltin*}(
```
Comment (juniset) [inline contracts/StructHash.cairo:70]: The other contracts of the repo use a different function declaration convention which I would suggest to follow for consistency (and readability).

```
func hashMessage{
        hash_ptr : HashBuiltin*
    } (
        domain : StarkNet_Domain,
        mail : Mail,
        account : felt
    ) -> (hash: felt):
```

File: test/struct_hash.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+@pytest.fixture(scope='module')
+async def get_starknet():
+    starknet = await Starknet.empty()
+    return starknet
+
+
+@pytest.mark.asyncio
+async def test_eip712(get_starknet):
```
Comment (juniset) [inline test/struct_hash.py:22]: Please no EIP712 :-)

Let's just call it "message signing".

File: contracts/StructHash.cairo (hunk 3 above)
Comment (janek26) [reply contracts/StructHash.cairo:70]: I'm using the auto formatter used by vscode-cairo plugin. Seems like the method definition is too short to go multiline

**Review by juniset — APPROVED** [review]

**Pull request merged**